src/engine/strat_engine/filesystem.rs

```diff
@@ -47,11 +56,25 @@ impl StratFilesystem {
         }
     }
 
-    pub fn check(&self, dm: &DM) -> EngineResult<FilesystemStatus> {
+    /// check if filesystem is getting full and needs to be extended
+    /// TODO: deal with the thindev in a Fail state.
+    pub fn check(&mut self, dm: &DM) -> EngineResult<FilesystemStatus> {
         match try!(self.thin_dev.status(dm)) {
-            ThinStatus::Good((_mapped, _highest)) => {
-                // TODO: check if filesystem is getting full and might need to
-                // be extended (hint: use statfs(2))
+            ThinStatus::Good(_) => {
+                let mount_point = try!(self.get_mount_point());
+                let (fs_total_bytes, fs_total_used_bytes) = try!(fs_usage(&mount_point));
+                let free_bytes = fs_total_bytes - fs_total_used_bytes;
+                if free_bytes.sectors() < FILESYSTEM_LOWATER {
+                    let extend_size = self.extend_size(self.thin_dev.size());
+                    match self.thin_dev.extend(dm, extend_size) {
+                        Ok(_) => {}
```

Comment:
Oops! One last thing! We need a call to `self.write_metadata` here, or else after reboot stratisd will not recognize the expanded thinpool data device.